pkg/controller/certificate-shim/sync.go

```diff
@@ -0,0 +1,598 @@
+
+		if err := translateAnnotations(crt, ingLike.GetAnnotations()); err != nil {
+			return nil, nil, err
+		}
+
+		// check if a Certificate for this TLS entry already exists, and if it
+		// does then skip this entry
+		if existingCrt != nil {
+			log := logf.WithRelatedResource(log, existingCrt)
+			log.V(logf.DebugLevel).Info("certificate already exists for this object, ensuring it is up to date")
+
+			if metav1.GetControllerOf(existingCrt) == nil {
+				log.V(logf.InfoLevel).Info("certificate resource has no owner. refusing to update non-owned certificate resource for object")
+				continue
+			}
+
+			if !metav1.IsControlledBy(existingCrt, ingLike) {
+				log.V(logf.InfoLevel).Info("certificate resource is not owned by this object. refusing to update non-owned certificate resource for object")
+				continue
+			}
+
+			if !certNeedsUpdate(existingCrt, crt) {
+				log.V(logf.DebugLevel).Info("certificate resource is already up to date for object")
+				continue
+			}
+
+			updateCrt := existingCrt.DeepCopy()
+
+			updateCrt.Spec = crt.Spec
+			updateCrt.Labels = crt.Labels
+
+			setIssuerSpecificConfig(crt, ingLike)
+
+			updateCrts = append(updateCrts, updateCrt)
+		} else {
+
+			newCrts = append(newCrts, crt)
+		}
+	}
+	return newCrts, updateCrts, nil
+}
+
+func findUnrequiredCertificates(list cmlisters.CertificateLister, ingLike metav1.Object) ([]*cmapi.Certificate, error) {
+	crts, err := list.Certificates(ingLike.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	var unrequired []*cmapi.Certificate
+	for _, crt := range crts {
+		if isUnrequiredCertificate(crt, ingLike) {
+			unrequired = append(unrequired, crt)
+		}
+	}
+
+	return unrequired, nil
+}
+
+func isUnrequiredCertificate(crt *cmapi.Certificate, ingLike metav1.Object) bool {
+	if !metav1.IsControlledBy(crt, ingLike) {
+		return false
+	}
+
+	switch o := ingLike.(type) {
+	case *networkingv1beta1.Ingress:
+		for _, tls := range o.Spec.TLS {
+			if crt.Spec.SecretName == tls.SecretName {
+				return false
+			}
+		}
+	case *gwapi.Gateway:
+		for _, l := range o.Spec.Listeners {
+			if crt.Spec.SecretName == l.TLS.CertificateRef.Name {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// certNeedsUpdate checks and returns true if two Certificates differ.
+func certNeedsUpdate(a, b *cmapi.Certificate) bool {
+	if a.Name != b.Name {
+		return true
+	}
+
+	// TODO: we may need to allow users to edit the managed Certificate resources
+	// to add their own labels directly.
+	// Right now, we'll reset/remove the label values back automatically.
+	// Let's hope no other controllers do this automatically, else we'll start fighting...
+	if !reflect.DeepEqual(a.Labels, b.Labels) {
+		return true
+	}
+
+	if a.Spec.CommonName != b.Spec.CommonName {
+		return true
+	}
+
+	if len(a.Spec.DNSNames) != len(b.Spec.DNSNames) {
+		return true
+	}
+
+	for i := range a.Spec.DNSNames {
+		if a.Spec.DNSNames[i] != b.Spec.DNSNames[i] {
```

Comment:
Note: I wonder if the order of the DNS names can ever change.

Reply:
yes, DNSNames might probably be considered as a "set", but I imagine that anyone would not be surprised to see their Certificate recreated if they swap the order of the DNS names. I'd leave it as is since it does not seem surprising.

Reply:
Agree that if user was to change it, it'd not be surprising if it was re-creating. I was wondering if there is a chance that we change the ordering in cert-manager somehow as we read them from the tls block, but assume that's not the case.